### PR TITLE
[HttpFoundation] parse cookie values containing the equal sign

### LINKED ISSUE #33667: Cookie::fromString() HeaderUtils::split() - Symfony's HttpFoundation
**Description**
Issue was identified when adding a Set-Cookie header with a query string value to a Response object, and was traced back to HeaderUtils::split() called from Cookie::fromString().

This function fails to parse cookie values with '=' symbols on it, which is perfectly fine according the RFC 6265.

 A string like "foo_cookie=foo=1&bar=2&baz=3" ends up being just "foo" in the Cookie object.

> 
> https://tools.ietf.org/html/rfc6265#section-4.1.1
> 
> cookie-value      = *cookie-octet / ( DQUOTE *cookie-octet DQUOTE )
>  cookie-octet      = %x21 / %x23-2B / %x2D-3A / %x3C-5B / %x5D-7E
>                        ; US-ASCII characters excluding CTLs,
>                        ; whitespace DQUOTE, comma, semicolon,
>                        ; and backslash

**How to reproduce**  

```php
	use Symfony\Component\HttpFoundation\Cookie;
    
	$header = "Set-Cookie: bjt_track=firstVisit=2019-09-23 06:18:41&sessionsCount=14&previousVisit=2019-09-23 06:26:37&lastVisit=2019-09-23 06:27:09; expires=Tue, 22-Sep-2020 06:27:09 GMT; path=/";

	$parts = explode(": ", $header, 2); 

	var_dump(Cookie::fromString($parts[1]));

/* OUTPUT: 

object(Symfony\Component\HttpFoundation\Cookie)#3 (10) {
  ["name":protected]=>
  string(9) "bjt_track"
  ["value":protected]=>
  string(10) "firstVisit"    <--- Note the broken value here
  ["domain":protected]=>
  NULL
  ["expire":protected]=>
  int(1600756029)
  ["path":protected]=>
  string(1) "/"
  ["secure":protected]=>
  bool(false)
  ["httpOnly":protected]=>
  bool(false)
  ["raw":"Symfony\Component\HttpFoundation\Cookie":private]=>
  bool(true)
  ["sameSite":"Symfony\Component\HttpFoundation\Cookie":private]=>
  NULL
  ["secureDefault":"Symfony\Component\HttpFoundation\Cookie":private]=>
  bool(false)
}
*/

```

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #33667
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Review by derrabus — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/HttpFoundation/Cookie.php [hunk 1]
```diff
@@ -62,7 +62,15 @@ public static function fromString($cookie, $decode = false)
         $part = array_shift($parts);
 
         $name = $decode ? urldecode($part[0]) : $part[0];
-        $value = isset($part[1]) ? ($decode ? urldecode($part[1]) : $part[1]) : null;
+        $value = null;
+
+        if (isset($part[1])) {
```
Comment (jderusse) [inline src/Symfony/Component/HttpFoundation/Cookie.php]: What if the input is `foo==`? are we sure that `$part[1]` is an empty string? Or HeaderUtils replaced empty strings by null? Are we sure last empty part won't be trimmed?
Maybe we can add  test for this case?

Comment (xabbuh) [reply src/Symfony/Component/HttpFoundation/Cookie.php]: Currently, that example would be parsed to have `foo` as the cookie value. I am actually not sure what would be the desired value here though.

Comment (xabbuh) [reply src/Symfony/Component/HttpFoundation/Cookie.php]: I think I found a solution.

File: src/Symfony/Component/HttpFoundation/Cookie.php [hunk 2]
```diff
@@ -62,7 +62,11 @@ public static function fromString($cookie, $decode = false)
         $part = array_shift($parts);
 
         $name = $decode ? urldecode($part[0]) : $part[0];
-        $value = isset($part[1]) ? ($decode ? urldecode($part[1]) : $part[1]) : null;
+        $value = isset($cookie[\strlen($name) + 1]) ? substr($cookie, \strlen($name) + 1, strcspn($cookie, ';', \strlen($name) + 1)) : null;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Cookie.php]: Isn't this something that should be fixed in `HeaderUtils::split()`?

File: src/Symfony/Component/HttpFoundation/HeaderUtils.php [hunk 3]
```diff
@@ -211,12 +211,19 @@ private static function groupParts(array $matches, string $separators): array
...
-                $parts[] = self::groupParts($matches, $partSeparators);
+                $parts[] = self::groupParts($matches, $partSeparators, false);
             }
         } else {
             foreach ($partMatches as $matches) {
                 $parts[] = self::unquote($matches[0][0]);
             }
+
+            if (!$first && 2 < count($parts)) {
+                $parts = [
```
Comment (jderusse) [inline src/Symfony/Component/HttpFoundation/HeaderUtils.php:228]: I believe this implode should be moved before the `unquote`.
ie. `foo="a=b"`. 
+ the loop at line 217 use only `matxh[0][0]` don't we loose something here?

File: src/Symfony/Component/HttpFoundation/Cookie.php (hunk 2 above)
Comment (xabbuh) [reply src/Symfony/Component/HttpFoundation/Cookie.php]: moved the patch to the `HeaderUtils` class

File: src/Symfony/Component/HttpFoundation/HeaderUtils.php (hunk 3 above)
Comment (xabbuh) [reply src/Symfony/Component/HttpFoundation/HeaderUtils.php:228]: IMO this works as expected, but I added a test anyway

**Review by nicolas-grekas — APPROVED** [review]

**Review by jderusse — APPROVED** [review]

**Pull request merged**